File: span/project.py
from __future__ import annotations

import json
import os
from pathlib import Path
from typing import Any


class Project:
    def __init__(self, project_dir: str | Path) -> None:
        self.root = Path(project_dir).resolve()
        self.sheets_dir = self.root / "sheets"
        self.annotations_dir = self.root / "annotations"
        self.manifest_path = self.root / "manifest.json"
        self.annotations_dir.mkdir(parents=True, exist_ok=True)

    def list_sheet_files(self) -> list[Path]:
        return sorted(
            p for p in self.sheets_dir.glob("*.png") if p.is_file()
        )

    def annotation_path_for_sheet(self, sheet_name: str) -> Path:
        stem = Path(sheet_name).stem
        return self.annotations_dir / f"{stem}.annotations.json"

    def load_annotations(self, sheet_name: str) -> dict[str, Any]:
        path = self.annotation_path_for_sheet(sheet_name)
        if not path.exists():
            return {"image": sheet_name, "annotations": []}
        try:
            return json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            return {"image": sheet_name, "annotations": []}

    def save_annotations(
        self, sheet_name: str, payload: dict[str, Any]
    ) -> None:
        self.annotations_dir.mkdir(parents=True, exist_ok=True)
        path = self.annotation_path_for_sheet(sheet_name)
        tmp = path.with_suffix(path.suffix + ".tmp")
        tmp.write_text(
            json.dumps(payload, indent=2) + "\n", encoding="utf-8"
        )
        os.replace(tmp, path)

    def load_manifest(self) -> dict[str, dict[str, Any]]:
        if not self.manifest_path.exists():
            return {}
        try:
            data = json.loads(
                self.manifest_path.read_text(encoding="utf-8")
            )
            result: dict[str, dict[str, Any]] = {}
            for asset in data.get("assets", []):
                file_name = asset.get("file")
                if isinstance(file_name, str):
                    result[file_name] = asset
            return result
        except json.JSONDecodeError:
            return {}

    def sheet_exists(self, sheet_name: str) -> bool:
        return (self.sheets_dir / sheet_name).is_file()

    def sheet_path(self, sheet_name: str) -> Path:
        return self.sheets_dir / sheet_name

File: src/bun/project.ts
import { readdir, mkdir, rename } from "fs/promises";
import { join, parse as parsePath } from "path";
import type { Annotation, SheetWithAnnotations } from "../shared/rpc-schema";

export class Project {
	readonly root: string;
	readonly sheetsDir: string;
	readonly annotationsDir: string;
	readonly manifestPath: string;

	constructor(projectDir: string) {
		this.root = projectDir;
		this.sheetsDir = join(projectDir, "sheets");
		this.annotationsDir = join(projectDir, "annotations");
		this.manifestPath = join(projectDir, "manifest.json");
	}

	async ensureAnnotationsDir(): Promise<void> {
		await mkdir(this.annotationsDir, { recursive: true });
	}

	async listSheetFiles(): Promise<string[]> {
		try {
			const entries = await readdir(this.sheetsDir);
			return entries
				.filter((f) => f.toLowerCase().endsWith(".png"))
				.sort();
		} catch {
			return [];
		}
	}

	annotationPathForSheet(sheetName: string): string {
		const stem = parsePath(sheetName).name;
		return join(this.annotationsDir, `${stem}.annotations.json`);
	}

	async loadAnnotations(
		sheetName: string,
	): Promise<{ image: string; annotations: Annotation[] }> {
		const path = this.annotationPathForSheet(sheetName);
		try {
			const file = Bun.file(path);
			if (!(await file.exists())) {
				return { image: sheetName, annotations: [] };
			}
			return await file.json();
		} catch {
			return { image: sheetName, annotations: [] };
		}
	}

	async saveAnnotations(
		sheetName: string,
		annotations: Annotation[],
	): Promise<void> {
		await this.ensureAnnotationsDir();
		const path = this.annotationPathForSheet(sheetName);
		const payload = { image: sheetName, annotations };
		const tmpPath = path + ".tmp";
		await Bun.write(tmpPath, JSON.stringify(payload, null, 2) + "\n");
		await rename(tmpPath, path);
	}

	async loadManifest(): Promise<Map<string, Record<string, unknown>>> {
		const result = new Map<string, Record<string, unknown>>();
		try {
			const file = Bun.file(this.manifestPath);
			if (!(await file.exists())) return result;
			const data = await file.json();
			for (const asset of data.assets ?? []) {
				if (typeof asset.file === "string") {
					result.set(asset.file, asset);
				}
			}
		} catch {
			// ignore
		}
		return result;
	}

	async getSheetImageBase64(sheetName: string): Promise<string> {
		const path = join(this.sheetsDir, sheetName);
		const file = Bun.file(path);
		const buffer = await file.arrayBuffer();
		const base64 = Buffer.from(buffer).toString("base64");
		return `data:image/png;base64,${base64}`;
	}

	async loadProjectAnnotations(): Promise<SheetWithAnnotations[]> {
		const manifest = await this.loadManifest();
		const files = await this.listSheetFiles();
		const sheets: SheetWithAnnotations[] = [];

		for (const file of files) {
			const asset = manifest.get(file) ?? {};
			const { annotations } = await this.loadAnnotations(file);
			const annotationFile = parsePath(
				this.annotationPathForSheet(file),
			).base;

			sheets.push({
				file,
				name: (asset.name as string) ?? parsePath(file).name,
				imageUrl: "",
				annotationFile: `annotations/${annotationFile}`,
				annotations,
			});
		}

		return sheets;
	}
}
